### PR TITLE
uint16 for png encode Issue #213

### PR DESCRIPTION
**Pull request by Suhasnama**
Added uint16 for png encoded Image Feature

### CODE REVIEW

**Comment by ChanchalKumarMaji** [comment]
Also add tests in `image_feature_test.py`.

File: tensorflow_datasets/core/features/image_feature.py [hunk 1]
```diff
@@ -123,8 +135,8 @@ def _encode_image(self, np_image):
     """Returns np_image encoded as jpeg or png."""
     if not self._runner:
       self._runner = utils.TFGraphRunner()
-    if np_image.dtype != np.uint8:
-      raise ValueError('Image should be uint8. Detected: %s.' % np_image.dtype)
+    if np_image.dtype not in ACCEPTABLE_NP_DTYPES:
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/features/image_feature.py]: This should match self._dtype. Please use https://www.tensorflow.org/api_docs/python/tf/dtypes/DType#as_numpy_dtype instead. You can remove `ACCEPTABLE_NP_DTYPES ` afterwards

File: tensorflow_datasets/core/features/image_feature.py [hunk 2]
```diff
@@ -83,17 +86,26 @@ def __init__(self, shape=None, encoding_format=None):
...
+    self._dtype = None
 
     # Set and validate values
     self.set_encoding_format(encoding_format or 'png')
     self.set_shape(shape or (None, None, 3))
+    self.set_dtype(dtype or tf.uint8)
+
+  def set_dtype(self , dtype):
+    """ Update the dtype. """
+    if dtype not in ACCEPTABLE_DTYPES:
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/features/image_feature.py]: np.uin16 can only be used for png. Please also add a test for this. Also specify this in the `__init__` args section.

File: tensorflow_datasets/core/features/image_feature.py [hunk 3]
```diff
@@ -146,6 +158,9 @@ def encode_example(self, image_or_path_or_fobj):
 
   def decode_example(self, example):
     """Reconstruct the image from the tf example."""
+    if self._encoding_format is 'png' and self._dtype is tf.uint16:
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/features/image_feature.py]: It is good practice to always reuse code whenever possible. Here for instance:
```python
 def decode_example(self, example):
  """Reconstruct the image from the tf example."""
  if self._encoding_format == 'png' and self._dtype == tf.uint16:
    decode_fn = tf.image.decode_png
  else:
    decode_fn = tf.image.decode_image
  img = decode_fn(example, channels=self._shape[-1], dtype=self._dtype)
  img.set_shape(self._shape)	    img.set_shape(self._shape)
  return img	    return img
```

Comment (Conchylicultor) [inline tensorflow_datasets/core/features/image_feature.py]: You should not use `is` when you want to test equality. The fact that it sometimes works due to the fact that Python is caching string make this behavior unpredictable.

See: https://stackoverflow.com/questions/1504717/why-does-comparing-strings-using-either-or-is-sometimes-produce-a-differe

**Comment by Conchylicultor** [comment]
Thank you for your contribution. I left a few comments

**Comment by Suhasnama** [comment]
Added the mentioned changes

File: tensorflow_datasets/core/features/image_feature.py (hunk 2 above)
Comment (Suhasnama) [reply tensorflow_datasets/core/features/image_feature.py]: @Conchylicultor  thanks for the reviewing and mentioning the changes , I made the requested changes , but couldn't add the test you mentioned (i.e. Test case to check np.uint16 is used only with 'png') , please help me with this. 
Thank you!!

Comment (Conchylicultor) [reply tensorflow_datasets/core/features/image_feature.py]: I don't understand. Why can't you add a new test here:
https://github.com/tensorflow/datasets/blob/master/tensorflow_datasets/core/features/image_feature_test.py

```python
self.assertFeature(
        feature=features_lib.Image(..., dtype=tf.uint16),
        shape=(None, None, 3),
        dtype=tf.uint16,
        tests=[
           ...
        ],
)
```

**Comment by Conchylicultor** [comment]
Thank you for your contribution.
It seems that the cla has expired. Also the diffs shows 300+ files modified. Could you rebase your brach to head to remove those diffs as it would create merge conflict (alternatively you can create a new PR if this is simpler).

**Comment by Suhasnama** [comment]
> Thank you for your contribution.
> It seems that the cla has expired. Also the diffs shows 300+ files modified. Could you rebase your brach to head to remove those diffs as it would create merge conflict (alternatively you can create a new PR if this is simpler).

Sure , will rebase my branch!!
Thank you

**Comment by Suhasnama** [comment]
Hello @Conchylicultor  , 
I have added "test_image_test" , please suggest me changes if any.
Thank you!!

File: tensorflow_datasets/core/features/image_feature_test.py [hunk 4]
```diff
@@ -54,7 +54,7 @@ def test_images(self):
             testing.FeatureExpectationItem(
                 value=img,
                 expected=img,
-            ),
+),
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/features/image_feature_test.py]: Merging artifact.

File: tensorflow_datasets/core/features/image_feature_test.py [hunk 5]
```diff
@@ -109,6 +109,22 @@ def test_image_shaped(self):
         ],
     )
 
+  def  test_image_dtype(self):
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/features/image_feature_test.py]: Please follow the style for consistency:
https://www.tensorflow.org/datasets/add_dataset#5_check_your_code_style

You can run pylint to ensure the file is compliant.
```
./oss_scripts/lint.sh this/file.py 
```

File: tensorflow_datasets/core/features/image_feature_test.py [hunk 6]
```diff
@@ -109,6 +109,22 @@ def test_image_shaped(self):
...
+      self.assertFeature(
+        feature=features_lib.Image( dtype=tf.uint16),
+        shape=(None, None, 3),
+        dtype=tf.uint8,
+        tests=[
+            testing.FeatureExpectationItem(
+                value = img , 
+                raise_cls = ValueError , 
+                raise_msg = 'dtype should be'
+            ),
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/features/image_feature_test.py]: It is good to test the error case, but you should also add a success case where the image is correctly decoded.

File: tensorflow_datasets/core/features/image_feature_test.py [hunk 7]
```diff
@@ -109,6 +109,22 @@ def test_image_shaped(self):
...
     )
 
+  def  test_image_dtype(self):
+      img = randint(256 , size=(128,128,3) ,dtype=np.uint8)
+      
+      # Image with statically defined dtype   
+      self.assertFeature(
+        feature=features_lib.Image( dtype=tf.uint16),
+        shape=(None, None, 3),
+        dtype=tf.uint8,
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/features/image_feature_test.py]: Why is output `dtype=tf.uint8` if the `Image` is set to `tf.uint16` ?
You should also modify the `get_tensor_info()`

Comment (Suhasnama) [reply tensorflow_datasets/core/features/image_feature_test.py]: Forgot to change dtype in `get_tensor_info()`  and  the output dtype also, now modified it! 
Thank you!!

File: tensorflow_datasets/core/features/image_feature_test.py (hunk 6 above)
Comment (Suhasnama) [reply tensorflow_datasets/core/features/image_feature_test.py]: Thanks for the review @Conchylicultor  , 
But when I run a new test case as following :
```
img_file_path = os.path.join(os.path.dirname(__file__),
                                 '../../testing/test_data/6pixels.png')
img_file_expected_content = [  # see tests_data/README.md
        [[0, 255, 0], [255, 0, 0], [255, 0, 255]],
        [[0, 0, 255], [255, 255, 0], [126, 127, 128]],
    ]
self.assertFeature(
        feature=features_lib.Image(dtype=tf.uint16),
        shape=(None, None, 3),
        dtype=tf.uint16,
        tests=[
            testing.FeatureExpectationItem(
                value = img_file_path , 
                expected = img_file_expected_content
            ),
        ],
    )
```
I was encountered with this error , couldn't figure out where did I go wrong .
```
AssertionError: 
Arrays are not equal

not equal where = (array([0, 0, 0, 0, 1, 1, 1, 1, 1, 1]), array([0, 1, 2, 2, 0, 1, 1, 2, 2, 2]), array([1, 0, 0, 2, 2, 0, 1, 0, 1, 2]))
not equal lhs = [65535 65535 65535 65535 65535 65535 65535 32382 32639 32896]
not equal rhs = [255 255 255 255 255 255 255 126 127 128]
Mismatch: 55.6%
Max absolute difference: 65280
Max relative difference: nan
 x: array([[[    0, 65535,     0],
        [65535,     0,     0],
        [65535,     0, 65535]],...
 y: array([[[  0, 255,   0],
        [255,   0,   0],
        [255,   0, 255]],...

```
Please help me!!

Comment (Conchylicultor) [reply tensorflow_datasets/core/features/image_feature_test.py]: That is what tests are for.

It seems that the image that you're trying to decode is encoded as int8 and you're trying to decode it as int16, so the decoding do not provide the correct result. You should instead create a new png image encoded as int16.

Additionally, you should also test when passing a np.ones(dtype=np.int16) directly that it is decoded correctly.

**Comment by lgeiger** [comment]
@Suhasnama @Conchylicultor Are there any updates on the state of the PR? This is a great addition and it seams like only a few small things are still missing

**Comment by lgeiger** [comment]
It looks like here `tf.uint8` is still hardcoded. I think this should be changed to `self._dtype`: https://github.com/tensorflow/datasets/pull/712/files#diff-a82ff144e5743da9160530df95d52bbcL116

**Comment by Conchylicultor** [comment]
There are still unresolved comments in this PR. Mainly, tests are missing. Don't hesitate to take over this PR to finish it, as I'm not sure when our team will have time to take over this.

**Comment by lgeiger** [comment]
> There are still unresolved comments in this PR. Mainly, tests are missing. Don't hesitate to take over this PR to finish it, as I'm not sure when our team will have time to take over this.

@Conchylicultor I made a new PR at #892

**Comment by Suhasnama** [comment]
I tried my best to complete this issue , but couldn't make it  .Thank you very much @Conchylicultor for helping me through out this issue , learnt a lot . Will try some other issue.
Thank you!

**Comment by Conchylicultor** [comment]
@Suhasnama no problem, thank you for starting this. @lgeiger finalized your works in https://github.com/tensorflow/datasets/pull/892 so it should be merged soon.
Thank you both for this work. Closing this PR.

**Pull request closed without merging**